File: frontend/src/api/interviews.ts
/* Interviews API client — the manual interview-loop log. Rounds are entered by
   hand (there is no calendar / ATS / email sync); the server returns the exact
   `Round` view-model with an inlined `debrief`, so the screen renders it
   unchanged. Every call is wrapped: reads fall back to an empty list, writes
   fall back to null so the UI can leave the row untouched on failure. */

export type RoundKind = "recruiter" | "hm" | "technical" | "onsite" | "values" | "other";
export type RoundStatus = "scheduled" | "done" | "cancelled";
export type DebriefSentiment = "good" | "mixed" | "tough" | "unknown";

export interface Debrief {
  sentiment: DebriefSentiment | null;
  notes: string | null;
  createdAt: string;
}

export interface Round {
  id: number;
  applicationId: number;
  kind: RoundKind | string;
  scheduledAt: string | null;
  status: RoundStatus | string;
  focus: string | null;
  debrief: Debrief | null;
}

export interface RoundCreate {
  kind: RoundKind | string;
  scheduledAt?: string | null;
  focus?: string | null;
  status?: RoundStatus | string;
}

export interface RoundPatch {
  kind?: RoundKind | string;
  scheduledAt?: string | null;
  focus?: string | null;
  status?: RoundStatus | string;
}

export interface DebriefCreate {
  sentiment?: DebriefSentiment | string;
  notes?: string | null;
}

const JSON_HEADERS = { "Content-Type": "application/json" };

async function getJSON<T>(url: string, fallback: T): Promise<T> {
  try {
    const r = await fetch(url);
    return r.ok ? ((await r.json()) as T) : fallback;
  } catch {
    return fallback;
  }
}

async function sendJSON(url: string, method: string, body?: unknown): Promise<Round | null> {
  try {
    const r = await fetch(url, {
      method,
      headers: JSON_HEADERS,
      body: body ? JSON.stringify(body) : undefined,
    });
    return r.ok ? ((await r.json()) as Round) : null;
  } catch {
    return null;
  }
}

export const listRounds = (appId: string) =>
  getJSON<Round[]>(`/api/applications/${appId}/rounds`, []);

export const createRound = (appId: string, body: RoundCreate) =>
  sendJSON(`/api/applications/${appId}/rounds`, "POST", body);

export const updateRound = (id: number, body: RoundPatch) =>
  sendJSON(`/api/rounds/${id}`, "PATCH", body);

export const captureDebrief = (id: number, body: DebriefCreate) =>
  sendJSON(`/api/rounds/${id}/debrief`, "POST", body);

export async function deleteRound(id: number): Promise<boolean> {
  try {
    const r = await fetch(`/api/rounds/${id}`, { method: "DELETE" });
    return r.ok;
  } catch {
    return false;
  }
}

File: src/matchbox/web/routes/interviews.py
"""JSON API for the interview loop (rounds + debriefs).

Powers the Workspace timeline. Rounds are manual entry (no calendar/ATS sync);
the debrief is a one-tap honest capture shown side-by-side with outcomes.
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from matchbox.interviews import repo
from matchbox.web.deps import ConnDep

router = APIRouter(prefix="/api")


class RoundCreateBody(BaseModel):
    kind: str
    scheduledAt: str | None = None
    focus: str | None = None
    status: str = "scheduled"


class RoundUpdateBody(BaseModel):
    kind: str | None = None
    scheduledAt: str | None = None
    focus: str | None = None
    status: str | None = None


class DebriefBody(BaseModel):
    sentiment: str | None = None
    notes: str | None = None


def _require(row: dict[str, Any] | None) -> dict[str, Any]:
    if row is None:
        raise HTTPException(status_code=404, detail="no such round")
    return row


@router.get("/applications/{app_id}/rounds")
def list_rounds(app_id: int, conn: ConnDep) -> list[dict[str, Any]]:
    return repo.rounds_for(conn, app_id)


@router.post("/applications/{app_id}/rounds")
def create_round(app_id: int, body: RoundCreateBody, conn: ConnDep) -> dict[str, Any]:
    try:
        rid = repo.create_round(
            conn,
            app_id,
            kind=body.kind,
            scheduled_at=body.scheduledAt,
            focus=body.focus,
            status=body.status,
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e
    return _require(repo.get_round(conn, rid))


@router.patch("/rounds/{round_id}")
def update_round(round_id: int, body: RoundUpdateBody, conn: ConnDep) -> dict[str, Any]:
    try:
        return _require(
            repo.update_round(
                conn,
                round_id,
                kind=body.kind,
                scheduled_at=body.scheduledAt,
                focus=body.focus,
                status=body.status,
            )
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e


@router.post("/rounds/{round_id}/debrief")
def capture_debrief(round_id: int, body: DebriefBody, conn: ConnDep) -> dict[str, Any]:
    if repo.get_round(conn, round_id) is None:
        raise HTTPException(status_code=404, detail="no such round")
    try:
        return _require(
            repo.upsert_debrief(conn, round_id, sentiment=body.sentiment, notes=body.notes)
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e


@router.delete("/rounds/{round_id}")
def delete_round(round_id: int, conn: ConnDep) -> dict[str, str]:
    repo.delete_round(conn, round_id)
    return {"ok": "true"}
